docs/langref/relay_type.rst

```diff
@@ -0,0 +1,213 @@
+===================
+Relay's Type System
+===================
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have not yet described the type system. Relay is
+a statically typed and type-inferred language, allowing programs to
+be typed with a minimal requirement of explicit type information.
+
+Static types are useful when performing compiler optimizations because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, and storage without needing to run the program. Additionally,
+static typing is useful for determining the 
+
+Relay's type system features a form of depending typing for shapes to
+replace the dynamic shape inference used in NNVM and most other machine
```

Comment:
Sorry. I didn't quite understand this (shouldn't the shape in NNVM is also static?). Could you please elaborate a bit? Thanks.

Reply:
I'll elaborate on it then. The idea is that shape inference is built into the type system and resolved entirely at compile time

Reply:
I see the point. I thought you were mentioning "dynamic shape". Thanks.